feat(common): 공통 예외 처리 시스템 구축

jiwonclvl opened the pull request:
## 📝 작업 내역
- [X] ErrorCode, CustomException, GlobalExceptionHandler를 도입하여 애플리케이션 전반의 에러를 일관되게 관리
- [X] 유효성 검사 실패 시, 어떤 필드에서 어떤 오류가 발생했는지 상세 정보를 담아 클라이언트에게 응답
- [X] @RestControllerAdvice가 처리하지 못하는 JWT 필터 예외를 필터 내에서 직접 처리
- [X] 예외 처리 관련 클래스들을 common/error 패키지로 재구성 및 에러 로그 포맷 개선

## 💡 PR 특이사항
JWT 필터에서 발생하는 예외 처리 로직에 대한 테스트는 아직 진행하지 못했습니다. 
추후 인증 토큰을 요구하는 API가 구현되면 해당 기능을 통해 테스트를 진행해야 할 것 같습니다. 

## 📸 스크린샷

### 이메일 중복 에러
<img width="938" height="416" alt="image" src="https://github.com/user-attachments/assets/501f1af7-fa7f-4f46-bc7f-a11baf66e926" />
<img width="910" height="123" alt="image" src="https://github.com/user-attachments/assets/d03e202a-5e0f-49c1-8a97-1f6fe208cc33" />

<br>

### validation 에러
<img width="943" height="608" alt="image" src="https://github.com/user-attachments/assets/84ca2329-3542-4886-98c3-6d35110e38be" />
<img width="1201" height="134" alt="image" src="https://github.com/user-attachments/assets/579efaa9-a506-4b47-946f-433a78834ba4" />

jiwonclvl committed:
feat(common): 공통 예외 처리 로직 추가

- 유효성 검사(Validation) 에러를 제외한 모든 애플리케이션 에러에 대해 HTTP 상태 코드와 메시지를 관리하는 ErrorCode enum을 정의했습니다.
- CustomException 클래스를 도입하여 @ExceptionHandler에서의 예외 처리 로직 중복을 최소화하고 일관된 처리를 가능하게 했습니다.
- @Valid어노테이션 사용을 위해 'spring-boot-starter-validation' 의존성을 추가했습니다.
- 클라이언트에게 일관된 형식의 에러 응답을 제공하기 위해 ErrorResponseDto를 생성했습니다.

jiwonclvl committed:
chore(error): 예외 처리 관련 파일 구조화 및 공통 예외 추가

jiwonclvl committed:
feat(error-handling): Validation 예외 상세 처리 및 로그 개선
- @Valid 어노테이션으로 발생하는 유효성 검사 실패에 대한 처리 로직을 추가했습니다.
- 각 유효성 검사 오류가 발생한 필드명과 메시지 ErrorDetailsDto에 담도록하였습니다.
- 에러 로그 형식을 텍스트 블록을 활용하여 가독성을 높이고 문제 파악을 용이하게 개선했습니다.
- ErrorResponseDto를 통합하여 모든 종류의 에러에 대해 일관된 응답 형식을 유지하도록 구조를 개선했습니다.

jiwonclvl committed:
feat(security): JWT 필터 예외 직접 처리 및 로그 포맷 개선

- Spring Security 필터에서 발생하는 JWT 관련 예외는 @RestControllerAdvice로 처리할 수 없으므로 필터 내에서 직접 예외를 처리하도록 구현했습니다.
- ObjectMapper를 사용하여 ErrorResponseDto를 JSON으로 직렬화하고, 이를 HttpServletResponse에 직접 작성하여 클라이언트에게 일관된 에러 응답을 제공하도록 수정했습니다.
- MethodArgumentNotValidException 발생 시, 여러 유효성 검사 오류가 포함된 List 객체가 로그에 배열 주소값이 아닌 포맷팅된 상세 내용으로 출력되도록 로그 메시지 형식을 개선했습니다.

JTPeg commented on src/main/java/piq/piqproject/common/error/dto/ErrorResponseDto.java:
```diff
@@ -4,17 +4,21 @@
         this.status = status;
         this.code = code;
         this.message = message;
+        this.errorDetails = errorDetails;
```
이렇게 하면 custom Exception발생시에 errorDetails가 null이 될 가능성이 존재해 보이는데 이에 대한 처리는 그냥 무시하는건지 확인 부탁

jiwonclvl replied:
실제로 customException의 경우 errorDetails가 항상 null이기 때문에 제네릭으로 동적으로 할당하거나 ValidationResponseDto를 별도로 만들어 처리할지 결정해야 할 것 같음

이 부분 의견 좀 듣고 싶음

jiwonclvl replied:
우선 DTO를 추가하여 구현하였는데 추후 더 나은 방법이 있다면 말해주면 좋을 것 같음

JTPeg commented on src/main/java/piq/piqproject/config/jwt/SecurityConfig.java:
```diff
@@ -82,7 +84,7 @@ public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
         // 다른 필터를 추가할 경우 여기에 추가할것
         // JWT Filter(custom Filter)를 Spring Security 이전에 추가
-        http.addFilterBefore(new JwtFilter(jwtTokenProvider), UsernamePasswordAuthenticationFilter.class);
+        http.addFilterBefore(new JwtFilter(jwtTokenProvider, objectMapper), UsernamePasswordAuthenticationFilter.class);
```
objectMapper를 Filter에서 생성하는게 더 나을것같았는데
그냥 bean으로 생성해서 넘겨주고 전역적으로 1개 objectMapper만 사용하도록 해야하는게 나을듯
나중에 objectMapper쓸일있으면 주입받고 사용하도록하고, objectMapper 설정하는 부분은 지금 없으면 나중에 추가해도좋을듯

jiwonclvl replied:
오 굳굳

jiwonclvl committed:
Merge branch 'develop' of https://github.com/PiqProject/piqProject into feat/common

# Conflicts:
#	build.gradle
#	src/main/java/piq/piqproject/config/jwt/JwtFilter.java

jiwonclvl committed:
refactor(exception): CustomException으로 예외 처리 통일

- 모든 비즈니스 로직 예외를 CustomException으로 수정하였습니다.
- super(message)를 적용하여 스택트레이스에 예외 메시지가 출력되도록 개선하였습니다.

jiwonclvl committed:
refactor(exception): Validation 에러 응답 DTO 추가

- CustomException 발생 시 errorDetails가 null인 문제를 해결하기 위해
  ValidationErrorResponseDTO를 통해 Validation 에러 필드를 별도로 처리하였습니다.
